chapters/interface.tex

```diff
@@ -540,6 +540,11 @@ \section{Function-Compatibility or Function-Subtyping for Functions}\label{funct
   constraining interface of the function being redeclared.
 \end{itemize}
 
+Note that function-compatibilty does currently not ensure that the function call has the correct variability, and that must then be checked after re-declaration.
```

Comment:
I find this kind of formulation quite controversial.  If we know our concepts are broken in this way, I think it makes more sense to have an open issue than to describe the limitation normatively.  If we should go the other way, the least we should do before adding a formulation like this – in my opinion – is to collect informed comments from all tool vendors.

Reply:
To me it makes sense to formulate the current limitations and later work on it separately - as explained above that it is not trivial - especially if we want to preserve compatibility for MSL.
However, if it helps I can remove that.

Reply:
The question is basically: how does tools handle models such as Modelica.Fluid.Machines.BaseClasses.PartialPump with a function-default having a parameter-value to ensure that the function call isn't used for a constant in the original model?

Since I haven't seen any previous complaints from any tool vendors about this construct I assume tools didn't do anything special and thus there is a "hole" in the check as stated in the new statement.

Reply:
@gkurzbach @sjoelund @eshmoylova consider Modelica.Fluid.Machines.ControlledPump and Modelica.Fluid.Machines.BaseClasses.PartialPump

And especially adding `final constant .Modelica.Units.SI.Position dzero = flowCharacteristic(0);` to PartialPump. 

Would you detect an error in PartialPump?
An error in the redeclare in the extends-clause of ControlledPump:
```
  redeclare replaceable function flowCharacteristic =
        Modelica.Fluid.Machines.BaseClasses.PumpCharacteristics.quadraticFlow (
          V_flow_nominal={0, V_flow_op, 1.5*V_flow_op},
          head_nominal={2*head_op, head_op, 0}))
```
Or some other special logic to detect this?
Alternatively is it just handled as part of normal variability checks (as suggested by this edit)?

Reply:
OpenModelica uses the normal variability checks and gives the following error:
```
Error: Component dzero of variability constant has binding 'HeatingSystem_total.pump.flowCharacteristic(0)' of higher variability parameter.
```

Reply:
> OpenModelica uses the normal variability checks and gives the following error:
> 
> ```
> Error: Component dzero of variability constant has binding 'flowCharacteristic(0)' of higher variability parameter.
> ```

And in case someone wonders: Dymola does the same (with a similar error message).

Reply:
> Since I haven't seen any previous complaints from any tool vendors about this construct I assume tools didn't do anything special and thus there is a "hole" in the check as stated in the new statement.

I can understand if tools are forced to open up holes because of shortcomings in the specification, but I'd like to avoid making this the prescribed way, as it could make it more difficult to develop a more solid variability logic in the future.

Reply:
> OpenModelica uses the normal variability checks and gives the following error:

Note that this discussion is about calling this _normal_ or not.  To me, it's not normal; the normal analysis would be that we have a function call with only constant arguments, hence it's a constant expression.  In other words, having parameter variability hiding in the (even replaceable!) function `flowCharacteristic` is _not_ normal procedure according to my reading of the current text.

Reply:
> > OpenModelica uses the normal variability checks and gives the following error:
> 
> Note that this discussion is about calling this _normal_ or not. 

But the proposed text doesn't say that it is _normal_, only that it hasn't been checked:
"Note that function-compatibilty does currently not ensure that the function call has the correct variability, and that must then be checked after re-declaration."

That just states that we _currently_ have missing guarantee here, and thus you need this extra check, so I don't see the problem with this text as a solution until we possibly find a better solution. I can, of course, remove this text or make it non-normative - but I don't see why we would be hung up on a word that isn't present in the text.

Reply:
The problem is that it is a normative statement telling that the way we define variability of function calls in 3.8 _Variability of Expressions_ is incorrect.

Reply:
> The problem is that it is a normative statement telling that the way we define variability of function calls in 3.8 _Variability of Expressions_ is incorrect.

No, it's saying that it's incomplete  - as these hidden arguments are also considered.
It could be clarified to say:
"Note that function-compatibilty does currently not ensure that the function call has the correct variability, and that must then be checked after re-declaration - by also considering the variability of these added arguments."

Reply:
In MapleSim we do the variability check on the arguments of `flowCharacteristic(0)` and do not check any any default arguments a function may have. So we do not detect the error for `dzero` in `ControlledPump`.

Reply:
In OpenModelica we first construct a function call with a complete argument list as specified in [12.4.1](https://specification.modelica.org/master/functions.html#positional-or-named-input-arguments-of-functions) and then we type check and determine the variability of the call. But I guess it's not specified that this is the way it must be done.

Reply:
I agree that there is a problem, but as I see it currently there are two possibilities - either you variability-check with the extra argument (as indicated by the new text; and already done by Dymola and OpenModelica) - or you check it earlier and thus you end up with equations breaking the variability-guarantees.

Obviously it would be better if this problem didn't exist - but until we have solved it I find it best to have it documented.